### PR TITLE
Cache AI Assistant welcome messages

### PR DESCRIPTION
**Pull request by fluiddot**
Related to 7851-gh-Automattic/dotcom-forge.

## Proposed Changes

- Refactor `useFetchWelcomeMessages` hook and use a React Context to expose and memorize welcome messages. The messages will be fetched when mounting the provider (this should happen only once per app session) and when the app gains focus. Additionally, when coming from an offline state, it will also fetch messages.
- Include welcome messages when the app is offline.
- Refactor content rendering of the assistant tab, especially the section where we render a notice.
- Add disabled state to welcome message prompts to cover offline case.

### Fetch welcome messages

- Add the following patch to log when welcome messages are fetched:
```patch
diff --git forkSrcPrefix/src/hooks/use-welcome-messages.tsx forkDstPrefix/src/hooks/use-welcome-messages.tsx
index 0482f1dfc114a2f7912fd2fbdb92e6cfe96a446e..b354556609aec00e84f25e4f6ac24351c757f53f 100644
--- forkSrcPrefix/src/hooks/use-welcome-messages.tsx
+++ forkDstPrefix/src/hooks/use-welcome-messages.tsx
@@ -42,6 +42,7 @@ export const WelcomeMessagesProvider = ( { children }: { children: React.ReactNo
 		}
 		isFetchingMessages.current = true;
 		try {
+			console.log( 'Fetching welcome messages' );
 			const response = await client.req.get( {
 				path: '/studio-app/ai-assistant/welcome',
 				apiNamespace: 'wpcom/v2',

```
- Run the app with the command `STUDIO_AI=true npm start`
- Open the Developer Tools.
- Observe that welcome messages are fetched upon opening.
- Navigate to the assistant tab.
- Observe the welcome messages are displayed.
- Select other sites.
- Observe the welcome messages are not fetched.
- Switch to another app and focus back on the Studio app.
- Observe that welcome messages are fetched upon opening.

> [!NOTE]
> Native modals in the app also trigger the focus event when they are dismissed.

### Assistant's tab renders the different states properly

**Not authenticated / Online:**

- Log out from WordPress.com.
- Navigate to the assistant tab.
- Observe the Hold up! message is displayed prompting to log in.
- Observe that no messages can be sent.

**Not authenticated / Offline:**

- Log out from WordPress.com.
- Turn off your internet connection.
- Navigate to the assistant tab.
- Observe the Hold up! message is displayed prompting to log in.
- Observe the offline notice is displayed.
- Observe that no messages can be sent.

**Authenticated / With no messages / Online:**

- Log in to WordPress.com.
- Navigate to the assistant tab.
- Clear conversation if it's not empty.
- Observe welcome messages are displayed.
- Click one of the example prompts.
- Observe that the AI responds.

**Authenticated / With no messages / Offline:**

- Log in to WordPress.com.
- Turn off your internet connection.
- Navigate to the assistant tab.
- Clear conversation if it's not empty.
- Observe welcome messages are displayed.
- Observe example prompts are disabled.
- Observe the offline notice is displayed.
- Observe that no messages can be sent.

**Authenticated / With messages / Offline:**

- Log in to WordPress.com.
- Navigate to the assistant tab.
- Clear conversation if it's not empty.
- Observe welcome messages are displayed.
- Click one of the example prompts.
- Observe that the AI responds.
- Turn off your internet connection.
- Observe that sent messages are displayed.
- Observe the offline notice is displayed.
- Observe that no messages can be sent.

**Authenticated / With messages / Online / Prompt limit reached:**

- Log in to WordPress.com.
- Navigate to the assistant tab.
- Clear conversation if it's not empty.
- Observe welcome messages are displayed.
- Click one of the example prompts.
- Observe that the AI responds.
- Apply the following patch:
```patch
diff --git forkSrcPrefix/src/hooks/use-prompt-usage.tsx forkDstPrefix/src/hooks/use-prompt-usage.tsx
index 752a589758b197b9140c9d12e1a9c6653741080f..fea9606b7cceabe44832b32021631361a4c5a488 100644
--- forkSrcPrefix/src/hooks/use-prompt-usage.tsx
+++ forkDstPrefix/src/hooks/use-prompt-usage.tsx
@@ -58,7 +58,7 @@ export function PromptUsageProvider( { children }: PromptUsageProps ) {
 				return;
 			}
 			setPromptLimit( limit );
-			setPromptCount( limit - remaining );
+			setPromptCount( limit );
 			if ( ! initiated ) {
 				setInitiated( true );
 			}

```
- Observe the usage limit notice is displayed.
- Observe that no messages can be sent.

**Authenticated / With messages / Online / Clear history threshold reached:**

- Log in to WordPress.com.
- Navigate to the assistant tab.
- Clear conversation if it's not empty.
- Observe welcome messages are displayed.
- Click one of the example prompts.
- Observe that the AI responds.
- Apply the following patch:
```patch
diff --git forkSrcPrefix/src/components/ai-clear-history-reminder.tsx forkDstPrefix/src/components/ai-clear-history-reminder.tsx
index 014501e30d3b76f9bd751827ac42428f177e0c24..02fe83ff8a14c85cd5782a972573038e607e2482 100644
--- forkSrcPrefix/src/components/ai-clear-history-reminder.tsx
+++ forkDstPrefix/src/components/ai-clear-history-reminder.tsx
@@ -77,9 +77,9 @@ function AIClearHistoryReminder( {
 		}
 	}, [ clearInput ] );
 
-	if ( ! showReminder ) {
-		return null;
-	}
+	// if ( ! showReminder ) {
+	// 	return null;
+	// }
 
 	return (
 		<div ref={ elementRef } className="mt-8 mb-2 text-center">

```
- Observe the clear history notice is displayed.
- Observe that messages can be sent.

**Authenticated / With messages / Offline / Prompt limit reached / Clear history threshold reached:**

- Log in to WordPress.com.
- Navigate to the assistant tab.
- Clear conversation if it's not empty.
- Observe welcome messages are displayed.
- Click one of the example prompts.
- Observe that the AI responds.
- Apply the following patch:
```patch
diff --git forkSrcPrefix/src/components/ai-clear-history-reminder.tsx forkDstPrefix/src/components/ai-clear-history-reminder.tsx
index 014501e30d3b76f9bd751827ac42428f177e0c24..02fe83ff8a14c85cd5782a972573038e607e2482 100644
--- forkSrcPrefix/src/components/ai-clear-history-reminder.tsx
+++ forkDstPrefix/src/components/ai-clear-history-reminder.tsx
@@ -77,9 +77,9 @@ function AIClearHistoryReminder( {
 		}
 	}, [ clearInput ] );
 
-	if ( ! showReminder ) {
-		return null;
-	}
+	// if ( ! showReminder ) {
+	// 	return null;
+	// }
 
 	return (
 		<div ref={ elementRef } className="mt-8 mb-2 text-center">
diff --git forkSrcPrefix/src/hooks/use-prompt-usage.tsx forkDstPrefix/src/hooks/use-prompt-usage.tsx
index 752a589758b197b9140c9d12e1a9c6653741080f..fea9606b7cceabe44832b32021631361a4c5a488 100644
--- forkSrcPrefix/src/hooks/use-prompt-usage.tsx
+++ forkDstPrefix/src/hooks/use-prompt-usage.tsx
@@ -58,7 +58,7 @@ export function PromptUsageProvider( { children }: PromptUsageProps ) {
 				return;
 			}
 			setPromptLimit( limit );
-			setPromptCount( limit - remaining );
+			setPromptCount( limit );
 			if ( ! initiated ) {
 				setInitiated( true );
 			}

```
- Observe the usage limit notice is displayed.
- Observe the clear history notice is not displayed.
- Observe that no messages can be sent.

**Authenticated / With messages / Online / Prompt limit reached / Clear history threshold reached:**

- Log in to WordPress.com.
- Navigate to the assistant tab.
- Clear conversation if it's not empty.
- Observe welcome messages are displayed.
- Click one of the example prompts.
- Observe that the AI responds.
- Apply the following patch:
```patch
diff --git forkSrcPrefix/src/components/ai-clear-history-reminder.tsx forkDstPrefix/src/components/ai-clear-history-reminder.tsx
index 014501e30d3b76f9bd751827ac42428f177e0c24..02fe83ff8a14c85cd5782a972573038e607e2482 100644
--- forkSrcPrefix/src/components/ai-clear-history-reminder.tsx
+++ forkDstPrefix/src/components/ai-clear-history-reminder.tsx
@@ -77,9 +77,9 @@ function AIClearHistoryReminder( {
 		}
 	}, [ clearInput ] );
 
-	if ( ! showReminder ) {
-		return null;
-	}
+	// if ( ! showReminder ) {
+	// 	return null;
+	// }
 
 	return (
 		<div ref={ elementRef } className="mt-8 mb-2 text-center">
diff --git forkSrcPrefix/src/hooks/use-prompt-usage.tsx forkDstPrefix/src/hooks/use-prompt-usage.tsx
index 752a589758b197b9140c9d12e1a9c6653741080f..fea9606b7cceabe44832b32021631361a4c5a488 100644
--- forkSrcPrefix/src/hooks/use-prompt-usage.tsx
+++ forkDstPrefix/src/hooks/use-prompt-usage.tsx
@@ -58,7 +58,7 @@ export function PromptUsageProvider( { children }: PromptUsageProps ) {
 				return;
 			}
 			setPromptLimit( limit );
-			setPromptCount( limit - remaining );
+			setPromptCount( limit );
 			if ( ! initiated ) {
 				setInitiated( true );
 			}

```
- Turn off your internet connection.
- Observe the offline notice is displayed.
- Observe the usage limit notice is not displayed.
- Observe the clear history notice is not displayed.
- Observe that no messages can be sent.

## Pre-merge Checklist

- [x] Have you checked for TypeScript, React or other console errors?

### CODE REVIEW

File: src/components/content-tab-assistant.tsx [hunk 1]
```diff
@@ -91,6 +91,10 @@ const AuthenticatedView = memo(
 			return () => clearTimeout( timer );
 		}, [ messages?.length ] );
 
+		if ( messages.length === 0 ) {
+			return null;
+		}
```
Comment (sejas) [inline src/components/content-tab-assistant.tsx:106]: 👍

File: src/components/site-content-tabs.tsx [hunk 2]
```diff
@@ -27,22 +28,24 @@ export function SiteContentTabs() {
...
+							{ name === 'overview' && <ContentTabOverview selectedSite={ selectedSite } /> }
+							{ name === 'share' && <ContentTabSnapshots selectedSite={ selectedSite } /> }
+							{ name === 'settings' && <ContentTabSettings selectedSite={ selectedSite } /> }
+							{ assistantEnabled && name === 'assistant' && (
+								<ContentTabAssistant selectedSite={ selectedSite } />
+							) }
+						</div>
+					) }
+				</TabPanel>
+			</WelcomeMessagesProvider>
```
Comment (sejas) [inline src/components/site-content-tabs.tsx]: Not a strong opinion, but what do you think about moving the provider to wrap only the ContentTabAssistant ? 

```diff
-		<WelcomeMessagesProvider>
			<TabPanel
				className="mt-6 h-full flex flex-col overflow-hidden"
				tabs={ tabs }
				orientation="horizontal"
			>
				{ ( { name } ) => (
					<div className="h-full">
						{ name === 'overview' && <ContentTabOverview selectedSite={ selectedSite } /> }
						{ name === 'share' && <ContentTabSnapshots selectedSite={ selectedSite } /> }
						{ name === 'settings' && <ContentTabSettings selectedSite={ selectedSite } /> }
+						<WelcomeMessagesProvider>
							{ assistantEnabled && name === 'assistant' && (
								<ContentTabAssistant selectedSite={ selectedSite } />
							) }
+						</WelcomeMessagesProvider>
					</div>
				) }
			</TabPanel>
-		</WelcomeMessagesProvider>
```

**Review by sejas — APPROVED** [review]
Looks great!
I tested it by switching sites and tabs.


https://github.com/Automattic/studio/assets/779993/73bb4b33-cc30-4977-8686-379c819d2153

Also tested in these scenarios:

auth-no-messages
auth-with-messages
unauthenticated

auth-offline


unauth-offline

File: src/components/site-content-tabs.tsx (hunk 2 above)
Comment (fluiddot) [reply src/components/site-content-tabs.tsx]: I originally placed this component at this location but realized it was being rendered on every navigation to the Assistant tab. This is the main reason why I moved it to an upper rendering level.

**Pull request merged**